### PR TITLE
Improve delete performance by reducing the number of calls

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
Improve delete performance by reducing the number of calls

## Motivation and Context
- Remove the requirement to honor storage class for deletes
- Improve `posix.DeleteFileBulk` code to Stat the volumeDir
  only once per call, rather than for all object paths.

## How to test this PR?
No functionality changes, all changes should work as is.

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation needed
- [ ] Unit tests needed
- [ ] Functional tests needed (If yes, add [mint](https://github.com/minio/mint) PR # here: )

### CODE REVIEW

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-worm.sh | :x: [more...](#c0eec592-cd9d-42b5-8f6a-bffca1938bb6) |
| mint-gateway-azure.sh | :x: [more...](#f6a390c0-abbd-4411-b858-7cbced55e261) |
<a name="c0eec592-cd9d-42b5-8f6a-bffca1938bb6"></a>
### 9092-e0700ca/mint-worm.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.55:32333
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            worm
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp c20c3780bd4c:/mint/log /tmp/mint-logs'

(1/1) Running worm tests ... FAILED in 2 minutes and 0 seconds
{
  "alert": "",
  "args": {
    "bucketName": "bucket-qh3pvm1wbf4w09h7j23m2sghfnk3m7bxr69e9yhq7twkrkm7gq4igl0",
    "expiry": 60000000000,
    "objectName": "testObject"
  },
  "duration": 120306,
  "error": "RequestError: send request failed\ncaused by: Put http://72.28.97.55:32333/bucket-qh3pvm1wbf4w09h7j23m2sghfnk3m7bxr69e9yhq7twkrkm7gq4igl0: dial tcp 72.28.97.55:32333: i/o timeout",
  "function": "PutAndDelete",
  "message": "WORM_MODE ON  - CreateBucket Failed",
  "name": "test worm mode",
  "status": "FAIL"
}

Executed 0 out of 1 tests successfully.
```
<a name="f6a390c0-abbd-4411-b858-7cbced55e261"></a>
### 9092-e0700ca/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.61:31885
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp e972c3d7347c:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 8 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 5 minutes and 39 seconds
(4/14) Running aws-sdk-ruby tests ... done in 17 seconds
(5/14) Running awscli tests ... done in 3 minutes and 17 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 7 minutes and 24 seconds
(8/14) Running minio-dotnet tests ... done in 1 minutes and 35 seconds
(9/14) Running minio-go tests ... done in 22 minutes and 35 seconds
(10/14) Running minio-java tests ... FAILED in 8 minutes and 56 seconds
{
  "name": "minio-java",
  "function": "composeObject(String bucketName, String objectName,List<ComposeSource> composeSources, Map <String,String > headerMap, ServerSideEncryption sseTarget)",
  "args": "size: 6 MB & 6 MB",
  "duration": 25048,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code=NotImplemented, message=A header you provided implies functionality that is not implemented, bucketName=minio-java-test-se65qo, objectName=minio-java-test-2kjao63, resource=/minio-java-test-se65qo/minio-java-test-2kjao63, requestId=15F98560C5BB4B2F, hostId=d49c0dc8-8474-430b-8351-22b902a0f3fd)\nrequest={method=PUT, url=http://72.28.97.61:31885/minio-java-test-se65qo/minio-java-test-2kjao63?uploadId=fb5b3c5f2b9cebbe&partNumber=1, headers=x-amz-copy-source-if-match: a2a6dab7956f6ef753f5f6684249a851-2\nx-amz-copy-source: minio-java-test-se65qo/minio-java-test-1lvdi5s\nAccept-Encoding: identity\nHost: 72.28.97.61:31885\nUser-Agent: MinIO (amd64; amd64) minio-java/dev\nx-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\nx-amz-date: 20200305T212314Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20200305/us-east-1/s3/aws4_request, SignedHeaders=accept-encoding;host;x-amz-content-sha256;x-amz-copy-source;x-amz-copy-source-if-match;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=501, headers=Accept-Ranges: bytes\nContent-Length: 408\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nServer: MinIO/DEVELOPMENT.2020-03-05T20-14-43Z\nVary: Origin\nX-Amz-Request-Id: 15F98560C5BB4B2F\nX-Xss-Protection: 1; mode=block\nDate: Thu, 05 Mar 2020 21:23:14 GMT\n}\n >>> [io.minio.MinioClient.executeReq(MinioClient.java:1220), io.minio.MinioClient.execute(MinioClient.java:1082), io.minio.MinioClient.executePut(MinioClient.java:1446), io.minio.MinioClient.executePut(MinioClient.java:1468), io.minio.MinioClient.uploadPartCopy(MinioClient.java:2619), io.minio.MinioClient.composeObject(MinioClient.java:2572), FunctionalTest.composeObject_test1(FunctionalTest.java:2587), FunctionalTest.runTests(FunctionalTest.java:3396), FunctionalTest.main(FunctionalTest.java:3514)]"
}

Executed 9 out of 14 tests successfully.
```
Deleting image on docker hub
Deleting image locally

File: cmd/posix.go [hunk 1]
```diff
@@ -1340,9 +1340,40 @@ func (s *posix) DeleteFile(volume, path string) (err error) {
...
+			return nil, errFaultyDisk
+		}
+		return nil, err
+	}
+
 	errs = make([]error, len(paths))
+	// Following code is needed so that we retain SlashSeparator
+	// suffix if any in path argument.
 	for idx, path := range paths {
-		errs[idx] = s.DeleteFile(volume, path)
```
Comment (vadmeste) [inline cmd/posix.go:1345]: Wondering why this change here, could you explain why you did this ?

File: cmd/xl-v1-object.go [hunk 2]
```diff
@@ -747,32 +748,43 @@ func (xl xlObjects) deleteObject(ctx context.Context, bucket, object string, wri
...
-					[]error{errFileNotFound, errFileAccessDenied})
-			} else {
-				disks, err = rename(ctx, disks, bucket, object, minioMetaTmpBucket, tmpObjs[i], true, writeQuorums[i],
-					[]error{errFileNotFound})
-			}
-			if err != nil {
-				errs[i] = err
-			}
+			wg.Add(1)
+			go func(idx int) {
```
Comment (vadmeste) [inline cmd/xl-v1-object.go]: Isn't too much to create a go-routine for each object ? especially if the server has a lot of other requests to process. We probably could have multiple workers (10, 20 ?) receiving the list of all objects

**Review by vadmeste — COMMENTED** [review]
Some comments

File: cmd/posix.go (hunk 1 above)
Comment (harshavardhana) [reply cmd/posix.go:1345]: It's the third item of improvement in the PR subject

File: cmd/xl-v1-object.go (hunk 2 above)
Comment (harshavardhana) [reply cmd/xl-v1-object.go]: Rename is quick, so it is not harmful to do this in parallel.

File: cmd/posix.go (hunk 1 above)
Comment (vadmeste) [reply cmd/posix.go:1345]: > It's the third item of improvement in the PR subject

oh right, I see

File: cmd/xl-v1-object.go (hunk 2 above)
Comment (vadmeste) [reply cmd/xl-v1-object.go]: > Rename is quick, so it is not harmful to do this in parallel.

Isn't the maximum number of goroutines is 10k ? what if we pass 10k objects to DeleteObjects() ?

Comment (harshavardhana) [reply cmd/xl-v1-object.go]: There is no maximum for goroutines, what is maximum is threads.

Comment (harshavardhana) [reply cmd/xl-v1-object.go]: I changed the code back to serial for now. As we have enough improvement already from the other two changes.

**Review by vadmeste — APPROVED** [review]
LGTM & tested

**Review by kannappanr — APPROVED** [review]
LGTM

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-fs.sh | :x: [more...](#5bfc7498-6536-4b98-90ed-1390f383e3da) |
| mint-dist-xl.sh | :x: [more...](#2056b62d-c867-48ef-af3a-bdf25bd3519a) |
| mint-gateway-azure.sh | :x: [more...](#df438ce4-4548-42b5-91b5-e5ed2adc2caf) |
<a name="5bfc7498-6536-4b98-90ed-1390f383e3da"></a>
### 9092-33f31fa/mint-fs.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.55:32645
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 6b68702d6fe7:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... FAILED in 2 minutes and 0 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-jgvbr4db4c4vg2",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 120345,
  "error": "RequestError: send request failed\ncaused by: Put http://72.28.97.55:32645/aws-sdk-go-test-jgvbr4db4c4vg2: dial tcp 72.28.97.55:32645: i/o timeout",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 14 tests successfully.
```
<a name="2056b62d-c867-48ef-af3a-bdf25bd3519a"></a>
### 9092-33f31fa/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.55:32508
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp a9476d87c6f5:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... FAILED in 2 minutes and 1 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-ly3gnhurc43tva",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 120364,
  "error": "RequestError: send request failed\ncaused by: Put http://72.28.97.55:32508/aws-sdk-go-test-ly3gnhurc43tva: dial tcp 72.28.97.55:32508: i/o timeout",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 14 tests successfully.
```
<a name="df438ce4-4548-42b5-91b5-e5ed2adc2caf"></a>
### 9092-33f31fa/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.52:31873
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 7fa0a928ee3f:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 9 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 5 minutes and 56 seconds
(4/14) Running aws-sdk-ruby tests ... done in 18 seconds
(5/14) Running awscli tests ... done in 2 minutes and 58 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 6 minutes and 1 seconds
(8/14) Running minio-dotnet tests ... done in 1 minutes and 34 seconds
(9/14) Running minio-go tests ... done in 19 minutes and 48 seconds
(10/14) Running minio-java tests ... FAILED in 8 minutes and 29 seconds
{
  "name": "minio-java",
  "function": "composeObject(String bucketName, String objectName,List<ComposeSource> composeSources, Map <String,String > headerMap, ServerSideEncryption sseTarget)",
  "args": "size: 6 MB & 6 MB",
  "duration": 11204,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code=NotImplemented, message=A header you provided implies functionality that is not implemented, bucketName=minio-java-test-31d9m6g, objectName=minio-java-test-3vvhoap, resource=/minio-java-test-31d9m6g/minio-java-test-3vvhoap, requestId=15F9DA53FFF10C64, hostId=313dbca3-6349-44b2-ba5a-7331b14eb184)\nrequest={method=PUT, url=http://72.28.97.52:31873/minio-java-test-31d9m6g/minio-java-test-3vvhoap?uploadId=fe647488c21e4d48&partNumber=1, headers=x-amz-copy-source-if-match: 135c96df636eda0535d9e1de0a002213-2\nx-amz-copy-source: minio-java-test-31d9m6g/minio-java-test-jbcfkv\nAccept-Encoding: identity\nHost: 72.28.97.52:31873\nUser-Agent: MinIO (amd64; amd64) minio-java/dev\nx-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\nx-amz-date: 20200306T231958Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20200306/us-east-1/s3/aws4_request, SignedHeaders=accept-encoding;host;x-amz-content-sha256;x-amz-copy-source;x-amz-copy-source-if-match;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=501, headers=Accept-Ranges: bytes\nContent-Length: 410\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nServer: MinIO/DEVELOPMENT.2020-03-06T22-15-32Z\nVary: Origin\nX-Amz-Request-Id: 15F9DA53FFF10C64\nX-Xss-Protection: 1; mode=block\nDate: Fri, 06 Mar 2020 23:19:58 GMT\n}\n >>> [io.minio.MinioClient.executeReq(MinioClient.java:1220), io.minio.MinioClient.execute(MinioClient.java:1082), io.minio.MinioClient.executePut(MinioClient.java:1446), io.minio.MinioClient.executePut(MinioClient.java:1468), io.minio.MinioClient.uploadPartCopy(MinioClient.java:2619), io.minio.MinioClient.composeObject(MinioClient.java:2572), FunctionalTest.composeObject_test1(FunctionalTest.java:2587), FunctionalTest.runTests(FunctionalTest.java:3396), FunctionalTest.main(FunctionalTest.java:3514)]"
}

Executed 9 out of 14 tests successfully.
```
Deleting image on docker hub
Deleting image locally
